Pahomov_lab_tasks

vRedniy-LumbeRt opened the pull request:
Описание в Readme.txt

vRedniy-LumbeRt committed:
Complete_lab1_variants_tasks_and_test_module_trial

Добавил 3 варианта заданий для защит на первую лабу и пробный код парсинга тестов пользователя.

vRedniy-LumbeRt committed:
labs_full_complete_tasks

Все задания на защиту сделаны, продуманы с учётом балансировки. Ссылку приложил в Readme

dart-mih commented on Readme.txt:
Лучше перенести этот файл в папку проекта , а то получается, что для всего репозитория комментарии
И глобально ко всем текстовым файлам и pdf - переведите их md формат, в том числе условия заданий и формулы в Latex обернуть...

dart-mih commented on Pahomov_project/Lab_tasks/Lab_1/Var_1/task_1.pdf:
Не совсем понял, почему во втором примере теста используется функция divide(), а не foo(): это просто опечатка или какой-то смысл был заложен?

dart-mih commented on Pahomov_project/Lab_tasks/Lab_1/Var_1/task_2.pdf:
Ничего не говорится про ситуацию, когда удаление/дублирование/переворот невозможны:
@
4
ab@
например

Нужно как-то проговорить эту ситуацию в условии

dart-mih commented on Pahomov_project/Lab_tasks/Lab_1/Var_1/task_2.pdf:
Из формулировки задания не совсем понятно, откуда студентам брать реализацию стека... Честно говоря, даже я не понял:)
Просто если в тексте прописано, что все это хранить обязательно в стеке, то может показаться, что стек уже реализован - если это так, то его реализацию нужно привести в задании
Если стек не реализован, а его реализация обязательна, то нужно заострить на этом внимание и, возможно, даже попытаться проверить код, распарсив его. 
+ когда будете писать код для проверки работы студента, проверьте, что не было использовано каких-то дополнительных предреализованных стеков и запрет на их использования также нужно прописать текстом 

P.S. Может показаться, что вся эта "душная" возня с формулировками бесполезна... Для 80-90% студентов этого хватит, но неплохо процент людей не против воспользоваться лазейками в формулировках, а иногда и вовсе возникают конфликтные ситуации на этом фоне

dart-mih commented on Pahomov_project/Lab_tasks/Lab_1/Var_1/task_1.pdf:
А еще забавно что task_1 имеет внутри номер 2, а task_2 - единицу соответственно:)

dart-mih commented on Pahomov_project/Testing module/Trying_to_understand_the_way_of_parsing_user_tests/for python/Py_check_tests.py:
```diff
@@ -0,0 +1,128 @@
+
+    # Запускаем все тесты и проверяем результат
+    if run_all_tests(file_path) != 0:
+        raise Exception('Tests Error: Не пройдены написанные пользователем тесты (проверьте корректность своих тестов).')
```
Когда будете переделывать проверку под формат мудла - осторожно с исключениями, их лучше отлавливать

dart-mih commented on Pahomov_project/Testing module/Trying_to_understand_the_way_of_parsing_user_tests/for python/Py_check_tests.py:
```diff
@@ -0,0 +1,128 @@
+                    for elem in mas:
+                        if isinstance(elem, ast.Call):  # Проверяем, есть ли после assert вызов функции
+                            # Проверяем, относится ли он к divide
+                            check_divide(elem)
```
С таким добротным анализом функций, на самом деле, можно попробовать сделать и проверку общей функциональности - но это уже по желанию вашему

dart-mih commented on Pahomov_project/Testing module/Readme.txt:
Глобально насчет модуля с тестами:
Признаюсь честно, я пока не запускал это дело, но идея хорошая. 
 - Единственное что сверху было бы неплохо еще сделать - если студент написал лишние тесты, то не стоит принимать такой ответ, вероятно. 
 - Нужно перенести все эти файлы в мудловский формат работы будет, что повлечет за собой некоторые изменения в коде: вероятно не будет файла с тестами, посколько код студента подается в текстовом формате и не будет такой структуры внутри одной сессии проверки... Если вам нужно поэкспериментировать с чем-то, а пока не создавать финальный тест для студентов, то можете без проблем создавать себе черновик для творчества на мудле (тесты можно удалять и редактировать, только не сотрите лишнего)
 
 P.S. Хотелось бы посмотреть, как это будет работать именно в формате когда код студента подается текстом, а не файлом

vRedniy-LumbeRt replied:
Здравствуйте! Касаемо тестов -- фактически, можно как ограничить, так и дать возможность пользователю тестировать всё, что ему вздумается.

В условиях задач будут просить именно граничные случаи отлавливать, но возможность писать и стандартные тесты, на мой взгляд, стоит оставить --- пользователь будет видеть, что его тест корректен, но явно не покрывает все случаи.

Оставляю итоговый вывод на Ваше усмотрение -- добавить подобную проверку -- дело нескольких минут

dart-mih commented on Pahomov_project/Testing module/Trying_to_understand_the_way_of_parsing_user_tests/for c++/C_check_tests.py:
```diff
@@ -0,0 +1,134 @@
+import clang.cindex  # Импортируем модуль clang для работы с AST
+
+# Устанавливаем путь к библиотеке libclang
+clang.cindex.Config.set_library_file("C:\\Users\\stepa\\PycharmProjects\\pythonProject11\\venv\\Lib\\site-packages\\clang\\native\\libclang.dll")  # замените на фактический путь к вашей libclang.dll
```
Ох не к добру это... Не проверяли, в мудле либа установлена?

vRedniy-LumbeRt replied:
Сегодня обращусь к Константину с просьбой проверить составляемый мною список библиотек (на созвоне было сказано, что сейчас идёт некая настройка, так что добавлять библиотеки хотел сам Константин)

dart-mih commented on Pahomov_project/Testing module/Trying_to_understand_the_way_of_parsing_user_tests/for c++/C_check_tests.py:
```diff
@@ -0,0 +1,134 @@
+    # Замените на путь к вашему файлу с тестами
+    _file_path = "C_tests.cpp"
+    # Получаем результаты анализа тестов
+    results = analyze_tests(_file_path)
```
Предлагаю послушать комментарий Константина из другого PR и перейти к классам:
1) не будет вложенных функций
2) можно сделать родительский класс общий и два дочерних под питон и плюсы, что сделает обращение при проверке на разных языках более удобным и читаемым
3) переменные между функциями можно будет просто заносить в поля класса - так удобнее их отслеживать

dart-mih commented on Pahomov_project/Testing module/Trying_to_understand_the_way_of_parsing_user_tests/for c++/C_check_tests.py:
```diff
@@ -0,0 +1,134 @@
+
+            # Проверяем значение аргумента и обновляем словарь test_cases
+            if value == 0:
+                test_cases["division_by_zero"] = True
```
Вынесите подобные константы строчные либо в самое начало файла в переменные, либо где-то в отдельном месте их определите... Они используются многократно и рефакторить будет не очень удобно

dart-mih commented on Pahomov_project/Testing module/Trying_to_understand_the_way_of_parsing_user_tests/for python/Py_check_tests.py:
```diff
@@ -0,0 +1,128 @@
+        raise Exception('Tests Error: Не пройдены написанные пользователем тесты (проверьте корректность своих тестов).')
+
+    # Анализируем тесты и выводим результаты
+    results = analyze_tests(file_path)
```
Аналогично про классы

dart-mih commented on Pahomov_project/Testing module/Trying_to_understand_the_way_of_parsing_user_tests/for python/Py_check_tests.py:
```diff
@@ -0,0 +1,128 @@
+
+    # Инициализация словаря для хранения результатов анализа
+    test_cases = {
+        "division_by_zero": False,
```
И про строчные константы

vRedniy-LumbeRt committed:
Create Lab_1 var_1 task_1.md

vRedniy-LumbeRt committed:
Delete Pahomov_project/Lab_tasks/Lab_1/Var_1/task_1.pdf

vRedniy-LumbeRt committed:
Delete Pahomov_project/Lab_tasks/Lab_1/Var_1/task_2.pdf

vRedniy-LumbeRt committed:
Update task_1.md

vRedniy-LumbeRt committed:
Update task_1.md

vRedniy-LumbeRt committed:
Delete Readme.txt

vRedniy-LumbeRt committed:
Delete Pahomov_project/Lab_tasks directory

vRedniy-LumbeRt committed:
Pahomov_IMPROVED_LABS

Исправил замечания по лабораторным работам, в соответствующей папке лежит Readme_NEW.
Следующий коммит сделаю для модуля тестирования.

vRedniy-LumbeRt committed:
Pahomov_IMPROVE_LABS_LAST

Почему-то слетело форматирование для примеров входа/выхода -- поправил

vRedniy-LumbeRt committed:
Delete Pahomov_project/Testing module directory

vRedniy-LumbeRt committed:
Pahomov_testing_module_working_model

Добавил готовую модель работы через классы для парсинга пользовательских тестов. Соответствующая информация в Readme папки тестов.

dart-mih commented on Pahomov_project/Lab_tasks/Lab_1/var_1.md:
```diff
@@ -0,0 +1,50 @@
+
+### **Задание 2**
+Протестировать работу функции foo, вычисляющей выражение 
+*y = 1 / sqrt(x^2 + 4x + 3)*
```
Тут описаны примеры, как формулы оформлять https://ashki23.github.io/markdown-latex.html

dart-mih commented on Pahomov_project/Lab_tasks/Lab_1/var_2.md:
```diff
@@ -0,0 +1,49 @@
+
+3
+
+abcde#nu#st
```
амперсант за решетку заменить надо

dart-mih commented on Pahomov_project/Lab_tasks/Lab_1/var_2.md:
````diff
@@ -0,0 +1,49 @@
+*F = (1 / n) + 1 / (n – 2)*
+
+
+```python
````
Рановато рамочку открыли
